src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
             }
         }
     }
+
+    private Void checkApplications() {
+        log.debug("Checking pending applications");
+        try {
+            long now = System.currentTimeMillis();
+
+            for (String redisKey : pendingApplications.valueRange(Double.NEGATIVE_INFINITY, true, now, true, 0, checkSize)) {
+                log.debug("Checking pending application: {}", redisKey);
+                ResourceDescription resource = ResourceDescription.fromAnyUrl(redisKey, encryptionService);
+
+                try {
+                    terminateApplication(resource, "Application failed to start in the specified interval");
+                } catch (Throwable e) {
+                    // ignore
+                }
+            }
+        } catch (Throwable e) {
+            log.warn("Failed to check pending applications:", e);
+        }
+
+        return null;
+    }
+
+    // region Launching Application
+
+    private Void launchApplication(ProxyContext context, ResourceDescription resource) {
+        try (LockService.Lock lock = lockService.tryLock(deploymentLockKey(resource))) {
```

Comment:
The operation may take a while. There is a timeout after that the lock will be released.

Reply:
This timeout is currently less than lock timeout. So I decided not to implement lock watchdog mechanism in this mr.

Reply:
Please add a comment about your decision was made.

Reply:
added